{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"19. Working with Exceptions.ipynb","version":"0.3.2","provenance":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"}},"cells":[{"metadata":{"id":"2mRK6hgdnWvJ","colab_type":"text"},"cell_type":"markdown","source":["### Exception Handling in Python\n","\n","- An exception is an error that happens during execution of a program"]},{"metadata":{"id":"izHT0PRmnWvU","colab_type":"code","colab":{}},"cell_type":"code","source":["In Python, exceptions can be handled using a try statement."],"execution_count":0,"outputs":[]},{"metadata":{"id":"TTNOFqxonWv2","colab_type":"text"},"cell_type":"markdown","source":["![Python%20Exceptions.jpg](attachment:Python%20Exceptions.jpg)"]},{"metadata":{"id":"ehZUQW-unWv9","colab_type":"text"},"cell_type":"markdown","source":["### Python Built-in Exceptions\n","\n","Exception::\t\t    Base class for all exceptions\n","\n","ArithmeticError:\tRaised when numeric calculations fails\n","\n","FloatingPointError:\tRaised when a floating point calculation fails\n","\n","ZeroDivisionError:\tRaised when division or modulo by zero takes place for all numeric types\n","\n","AssertionError:\t\tRaised when Assert statement fails\n","\n","OverflowError:\t\tRaised when result of an arithmetic operation is too large to be represented\n","\n","ImportError:\t\tRaised when the imported module is not found\n","\n","IndexError:\t\t    Raised when index of a sequence is out of range\n","\n","KeyboardInterrupt:\tRaised when the user interrupts program execution, generally by pressing Ctrl+c\n","\n","IndentationError:\tRaised when there is incorrect indentation\n","\n","SyntaxError:\t\tRaised by parser when syntax error is encountered\n","\n","KeyError:\t\t    Raised when the specified key is not found in the dictionary\n","\n","NameError:\t\t    Raised when an identifier is not found in the local or global namespace\n","\n","TypeError:\t\t    Raised when a function or operation is applied to an object of incorrect type\n","\n","ValueError:\t\t    Raised when a function gets argument of correct type but improper value\n","\n","IOError:\t\t    Raised when an input/ output operation fails\n","\n","RuntimeError:\t\tRaised when a generated error does not fall into any category"]},{"metadata":{"id":"ZTYPCPXwnWwL","colab_type":"code","colab":{}},"cell_type":"code","source":[""],"execution_count":0,"outputs":[]},{"metadata":{"id":"TdUTrNlHnWwd","colab_type":"code","colab":{},"outputId":"df32be91-89c7-4b8d-d657-69102cb077dc"},"cell_type":"code","source":["# import module sys to get the type of exception\n","import sys\n","\n","randomList = ['a', 0, 2]\n","\n","for entry in randomList:\n","    try:\n","        print(\"The entry is\", entry)\n","        r = 1/int(entry)\n","        break\n","    except:\n","        print(\"Oops!\",sys.exc_info()[0],\"occured.\")\n","        print(\"Next entry.\\n\")\n","print(\"The reciprocal of\",entry,\"is\",r)"],"execution_count":0,"outputs":[{"output_type":"stream","text":["The entry is a\n","Oops! <class 'ValueError'> occured.\n","Next entry.\n","\n","The entry is 0\n","Oops! <class 'ZeroDivisionError'> occured.\n","Next entry.\n","\n","The entry is 2\n","The reciprocal of 2 is 0.5\n"],"name":"stdout"}]},{"metadata":{"id":"_656mnR7nWxO","colab_type":"code","colab":{},"outputId":"195bf7e1-4d0e-40c2-d3eb-140ceedd29d1"},"cell_type":"code","source":["# Catching Specific Exceptions in Python\n","\n","import sys\n","\n","randomList = ['a', 0, 2]\n","\n","for entry in randomList:\n","    try:\n","        print(\"The entry is\", entry)\n","        r = 1/int(entry)\n","        break\n","    except ValueError:\n","        # handle ValueError exception\n","        print(\"Oops! vallue error occured.\")\n","    except (TypeError, ZeroDivisionError):\n","        # handle multiple exceptions\n","        # TypeError and ZeroDivisionError\n","        print(\"Oops! TypeError or  ZeroDivisionError has occured\")\n","    except:\n","        print(\"Oops!\",sys.exc_info()[0],\"occured.\")\n","        print(\"Next entry.\\n\")\n","print(\"The reciprocal of\",entry,\"is\",r)"],"execution_count":0,"outputs":[{"output_type":"stream","text":["The entry is a\n","Oops! vallue error occured.\n","The entry is 0\n","Oops! <class 'ZeroDivisionError'> occured.\n","Next entry.\n","\n","The entry is 2\n","The reciprocal of 2 is 0.5\n"],"name":"stdout"}]},{"metadata":{"id":"y_zQqj9FnWxr","colab_type":"code","colab":{}},"cell_type":"code","source":["Raising Exceptions\n","In Python programming, exceptions are raised when corresponding errors occur at run time, but we can forcefully raise it using the keyword raise."],"execution_count":0,"outputs":[]},{"metadata":{"id":"_61aAoBEnWyD","colab_type":"code","colab":{},"outputId":"bcc0f56c-a738-4c36-af51-d5af797896a1"},"cell_type":"code","source":["try:\n","    a = int(input(\"Enter a positive integer: \"))\n","    if a <= 0:\n","        raise ValueError(\"That is not a positive number!\")\n","except ValueError as ve:\n","    print(ve)"],"execution_count":0,"outputs":[{"output_type":"stream","text":["Enter a positive integer: -45\n","That is not a positive number!\n"],"name":"stdout"}]},{"metadata":{"id":"09Vy5nLenWye","colab_type":"code","colab":{}},"cell_type":"code","source":["## try...finally"],"execution_count":0,"outputs":[]},{"metadata":{"id":"sgzcW1gPnWy7","colab_type":"code","colab":{},"outputId":"4e4fbaf2-c12c-4780-ec49-7e916df1c1a2"},"cell_type":"code","source":["try:\n","    a = int(input(\"Enter a positive integer: \"))\n","    if a <= 0:\n","        raise ValueError(\"That is not a positive number!\")\n","except ValueError as ve:\n","    print(ve)\n","except:\n","    print(\"Unknown error occurred\")\n","else:\n","    print(\"No error is occurred\")\n","finally:\n","    print(\"Hello, I will always be there for your help\")"],"execution_count":0,"outputs":[{"output_type":"stream","text":["Enter a positive integer: 45\n","No error is occurred\n","Hello, I will always be there for your help\n"],"name":"stdout"}]},{"metadata":{"id":"Gw0fQrqpnWzZ","colab_type":"code","colab":{}},"cell_type":"code","source":[""],"execution_count":0,"outputs":[]},{"metadata":{"id":"awyM6-nVnWz7","colab_type":"code","colab":{}},"cell_type":"code","source":[""],"execution_count":0,"outputs":[]}]}